File: communicationsNetworking/Network/CRC.cpp
#include <iostream>
#include <string>

using namespace std;

string operator ^(string op1, string op2){
  if(op1.size() != op2.size()){
    cout<<"Error!"<<endl;
    exit(1);
  }

  string ans(op1.size()-1,'0');

  for(int i=0;i<op1.size()-1;i++){
    if(op1[i+1]==op2[i+1])
      ans[i]='0';
    else
      ans[i]='1';
  }

  return ans;
}


int main(){

  string data,divisor,temp,dividend;
  int k;

  cout<<"Enter the data: ";
  cin>>data;

  cout<<"Enter the divisor: ";
  cin>>divisor;

  k=divisor.size();
  dividend= data + string(k-1,'0');
  temp=dividend.substr(0,k);

  while(k<dividend.size()){
    if(temp[0]=='1')
      temp= (temp^divisor) + dividend[k];
    else
      temp= (temp^string(temp.size(),'0')) + dividend[k];
    k++;
  }
  if(temp[0]=='1')
    temp= temp^divisor;
  else
    temp= temp^string(temp.size(),'0');

  cout<<"Data to be sent by transmitter is : "<<data+temp<<endl;
  return 0;
}

File: communicationsNetworking/Network/CRC.py

def xor(op1,op2):
  result=[]
  for i in range(1,len(op2)):
      if op1[i] is op2[i]:
          result.append('0')
      else:
          result.append('1')

  return ''.join(result)

def remainder(dividend,divisor='1101'):
    k=len(divisor)
    temp=dividend[:k]

    while k<len(dividend):

        if temp[0] is '1':
            temp= xor(divisor,temp) + dividend[k]
        else:
            temp= xor('0'*k,temp) + dividend[k]

        k+=1
    if temp[0] is '1':
        temp= xor(divisor,temp)
    else:
        temp= xor('0'*k,temp)
    return temp

print("Enter the data to be sent: ")
data=str(input())
print("Enter the data recieved at the reciever side: ")
recieved=str(input())

div='1101'
data=data + remainder(data+'0'*(len(div)-1),div)
remain=remainder(recieved)

correct=True
for i in remain:
    if i is '1':
        correct=False
        break

if(correct):
    print("According to CRC the recieved data is correct")
else:
    print("The Data is Corrupt. Call for Retransmission. The data should have been ",data,".")
